sca: gcc: add support for additional analyzer options

AlexFabre opened the pull request:
This commit brings support for additional GCC static analyzer options with 'GCC_SCA_OPTS=...'

Linked to discussion on 'ccache' side effect on analyzer file generation[1]

[1] https://github.com/zephyrproject-rtos/zephyr/discussions/86196

AlexFabre committed:
sca: gcc: add support for additional analyzer options

This commit brings support for additional GCC static analyzer options
with 'GCC_SCA_OPTS=...'

Linked to discussion on 'ccache' side effect on analyzer file
generation[1]

[1] https://github.com/zephyrproject-rtos/zephyr/discussions/86196

Signed-off-by: Alex Fabre <alex.fabre@rtone.fr>

nordicjm commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
+if(DEFINED GCC_SCA_OPTS)
+    string(REPLACE "," ";" analyzer_option_list ${GCC_SCA_OPTS})
+    foreach(analyzer_option IN LISTS analyzer_option_list)
+
```
```suggestion
```
plus indent missing

AlexFabre replied:
Thanks Fixed

nordicjm commented on doc/develop/sca/gcc.rst:
```diff
@@ -16,3 +16,41 @@ called with a ``-DZEPHYR_SCA_VARIANT=gcc`` parameter, e.g.
+
+.. code-block:: shell
+
+    west build -b stm32h573i_dk samples/hello_world -- -DZEPHYR_SCA_VARIANT=gcc -DGCC_SCA_OPTS="-fdiagnostics-format=sarif-file,-fanalyzer-show-events-in-system-headers"
```
this does not work with sdk-ng 0.16.8:
```
arm-zephyr-eabi-gcc: error: unrecognized argument in option '-fdiagnostics-format=sarif-file'
arm-zephyr-eabi-gcc: note: valid arguments to '-fdiagnostics-format=' are: json text
```
so should be adjusted to something that does. If the first is changed to json then the error becomes:
```
[{"kind": "error", "column-origin": 1, "children": [], "escape-source": false, "locations": [], "message": "unrecognized command-line option '-fanalyzer-show-events-in-system-headers'"}]
```

AlexFabre replied:
You are right, I've written options that appeared with GCC 14.

Changing the example with options working with current zephyr-gcc (GCC 12).

tejlmand replied:
because sdk-ng is based on gcc 12, and you need gcc >=14 to support json.

AlexFabre replied:
Actually gcc 12 supports `json` but not the second option : `-fanalyzer-show-events-in-system-headers`

AlexFabre commented on doc/develop/sca/gcc.rst:
Fixed

AlexFabre commented on doc/develop/sca/gcc.rst:
```diff
@@ -16,3 +16,41 @@ called with a ``-DZEPHYR_SCA_VARIANT=gcc`` parameter, e.g.
+
+.. code-block:: shell
+
+    west build -b stm32h573i_dk samples/hello_world -- -DZEPHYR_SCA_VARIANT=gcc -DGCC_SCA_OPTS="-fdiagnostics-format=sarif-file,-fanalyzer-show-events-in-system-headers"
```
Thanks, fixed

AlexFabre commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
+        # Therefore, currently, the build process is halted if 'ccache' is enabled.
+        # [1] https://github.com/ccache/ccache/issues/1466
+        if(USE_CCACHE STREQUAL "0")
+        else()
```
Thanks, fixed

tejlmand replied:
actually `STREQUAL` should not be used.

Please just check for `if(NOT USE_CCACHE)`.

I will update ccache CMake module to properly use the normal CMake known constant check.

AlexFabre replied:
> Please just check for if(NOT USE_CCACHE).

That is not the same logic.

`USE_CCACHE=0` satisfies your `if` condition.

> actually STREQUAL should not be used.

Okay but this is currently the way `USE_CCACHE` is checked in the ccache module, so maybe it is not in the scope of that PR to change it.

AlexFabre commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
 
+# Get GCC version
+execute_process(
+    COMMAND ${CMAKE_C_COMPILER} -dumpfullversion
```
Thanks, fixed

tejlmand commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,45 @@
+  foreach(analyzer_option IN LISTS analyzer_option_list)
+
+    if(analyzer_option STREQUAL "-fdiagnostics-format=sarif-file" OR analyzer_option STREQUAL "-fdiagnostics-format=json-file")
+      # Beginning with GCC 14, users can use the -fdiagnostics-format option
+      # to output analyzer diagnostics as SARIF or JSON files.
+      # This option encounters a specific issue[1] when used with ccache.
+      # Therefore, currently, the build process is halted if 'ccache' is enabled.
+      # [1] https://github.com/ccache/ccache/issues/1466
```
and what to do when ccache fixes this issue and some users might have newer ccache installed while others are still stuck with the an old distro ccache ?

Perhaps a CMake warning is better in this case, instead of FATAL_ERROR.

Note, this is not a request for change, but opening the discussion to get opinions.

AlexFabre replied:
You are right, I've not figured out the best way to handle resolution.

Maybe asking for 'ccache' being disabled is not a bad way to remember this limitation over time ?

tejlmand replied:
it's not, but in current form we are guaranteed to get a but report raised by the first user who installs a ccache where this problem is fixed :cry: 

For time being, keeping current solution will bring best awareness to users trying the sarif-file output.

tejlmand reviewed (changes requested):
Thanks for this proposal.

A couple of details to address before approval.

AlexFabre commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
+# GCC static analyzer requires GCC version >= 10.0.0
+if("${temp_compiler_version}" VERSION_LESS 10.0.0)
+    message(FATAL_ERROR "GCC static analyzer requires GCC >= v10.0")
+endif()
```
`CMAKE_C_COMPILER_VERSION` is not yet resolved at this stage...

AlexFabre replied:
Ok, changed for `GCC_COMPILER_VERSION`

AlexFabre commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
+    message(FATAL_ERROR "GCC static analyzer requires GCC >= v10.0")
+endif()
+
+message(STATUS "Found SCA: GCC static analyzer ${temp_compiler_version}")
```
I just wanted to harmonize SCA tools logs.

I think it's a good thing to see in the logs that a SCA tool was called.

tejlmand replied:
but you're not calling a dedicated SCA tool, you're adding compile flags to the compiler.

If we want to print that an SCA tool is defined and used, then perhaps a generic message stating use of SCA and from where is better done here: https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindScaTools.cmake
so that it shows exactly the CMake SCA reference (which may be different from an actual tool).

Similar to how Zephyr toolchain variant is printed, and where the compiler of the toolchain can be named differently.
Toolchain: Zephyr, compiler: GNU, linker: ld
Toolchain: llvm, compiler: clang, linker: ld / lld

AlexFabre replied:
You are right, `[FindScaTools.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindScaTools.cmake)` is a better place for that king of message.

Removed the log.

AlexFabre commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,46 @@
+zephyr_get(USE_CCACHE)
+
+if(DEFINED GCC_SCA_OPTS)
+    string(REPLACE "," ";" analyzer_option_list ${GCC_SCA_OPTS})
```
I had stated in the documentation that user should input a "coma separated list".

I prefer what you suggest, so I changed the documentation for a "semicolon separated list" and simplified the parsing.

pdgendt commented on doc/develop/sca/gcc.rst:
```diff
@@ -13,6 +13,58 @@ Run GCC static analysis
+   :board: stm32h573i_dk
+   :gen-args: -DZEPHYR_SCA_VARIANT=gcc -DZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb -DGNUARMEMB_TOOLCHAIN_PATH=...
+   :goals: build
+   :compact:
```
A blank line is needed here, I think.

tejlmand commented on cmake/sca/gcc/sca.cmake:
```diff
@@ -1,5 +1,43 @@
+# Copyright (c) 2025 Alex Fabre
 
+# Get GCC version
+execute_process(
+  COMMAND ${CMAKE_C_COMPILER} -dumpfullversion
+  OUTPUT_VARIABLE GCC_COMPILER_VERSION
+  OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
```
The intention with the comment in https://github.com/zephyrproject-rtos/zephyr/pull/86290#discussion_r1971496939
was to remove this code completely, and instead have the compiler specific target code in https://github.com/zephyrproject-rtos/zephyr/blob/bc3a01a7e5c73a778b119717b56f673faa1e06b0/cmake/compiler/gcc/target.cmake#L42

This way the knowledge of how to obtain GCC version only needs to be maintained at one location (the compiler infrastructure).

To make things easy I made a commit with the intended changes here:
https://github.com/tejlmand/zephyr/commit/8af2e94708b13b3e54641e4c8fde820e8df95927

you may include https://github.com/tejlmand/zephyr/commit/8af2e94708b13b3e54641e4c8fde820e8df95927 in this PR.

```suggestion
```

tejlmand reviewed (commented):
Thanks.
Almost there.

I've made a small commit to assist on `GCC_COMPILER_VERSION`.

tejlmand committed:
cmake: introduce GCC_COMPILER_VERSION variable

Introduce GCC_COMPILER_VERSION to be used in Zephyr CMake when knowledge
of the current GCC Compiler version is needed and
CMAKE_C_COMPILER_VERSION has not been defined yet.

CMAKE_C_COMPILER_VERSION is not defined until project() is called.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

AlexFabre commented:
I completely messed up adding your commit @tejlmand, I let you double check if everything is good 🙈

AlexFabre commented:
Okay I have the CI complaining: `needs to match one of the signed-off-by entries`but compliance check works well on my machine. Any idea ?

tejlmand commented:
> Okay I have the CI complaining: `needs to match one of the signed-off-by entries`but compliance check works well on my machine. Any idea ?

because somehow the author changed when you cherry-picked my commit.
I have no idea how that happened in the process, but when I fetch this PR and look at the author, then I see:
```
$ git log -n1 ddca3deaef18b194cc972e901fd9f71613d8e790
commit ddca3deaef18b194cc972e901fd9f71613d8e790
Author: Torsten Tejlmand Rasmussen <torsten.rasmussen@nordicsemi.no>
Date:   Thu Feb 27 17:19:48 2025 +0100
    ...
    Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>
```
and as you see, the Author != Signed-off-by, as `Tejlmand` is not part of my signed off.

But when I look at my original commit 8af2e94708b13b3e54641e4c8fde820e8df95927, it looks like this:
```
$ git log -n1 8af2e94708b13b3e54641e4c8fde820e8df95927 
commit 8af2e94708b13b3e54641e4c8fde820e8df95927
Author: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
Date:   Thu Feb 27 16:26:58 2025 +0100
    ...
    Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
```
then the author and signed-off matches.

AlexFabre commented:
@nordicjm and @tejlmand I let you re-check
